include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+#pragma once
+
+#include <mutex>
+#include <shared_mutex>
+
+#include <rclcpp>
+
+namespace rosparam_shortcuts {
+
+/*
+ * Thread-safe container for generic parameter values, providing callback functions.
+ * This class implements a tree structure of parameters which are declared recursively.
+ * All leafs are of type Parameter<rclcpp::Parameter> which are used for interacting with
+ * the rclcpp node parameter interface using the NodeConfig class further down.
+ */
+template <typename ParameterT>
+class Parameter
+{
+public:
+	Parameter(const std::string& name);
+	Parameter(const std::string& name, const ParameterT& default_value);
+
+	// Fields
+	const std::string& name() const noexcept;
+	std::string description() const noexcept;
+	std::string additional_constraints() const noexcept;
+	bool isNodeParameter() const noexcept;  // true if this parameter contains an rclcpp::Parameter
+
+	// Properties
+	Parameter<ParameterT>& describe(const std::string& description, const std::string& additional_constraints = "");
+	// Allow setting additional properties. i.e. read_only, dynamic_typing?
+	// void setProperties(...);
+
+	// Validity Status
+	const bool isValid() const noexcept;  // errors empty?
+	const bool hasErrors() const noexcept;
+	const bool hasWarnings() const noexcept;
+	// TODO: Insert additional thread-safe access to errors/warnings
+
+	// Read/write-locked data access using std::shared_mutex
+	// See: https://www.youtube.com/watch?v=ozOgzlxIsdg
+	ReadLockedPtr<ParameterT> operator->const;
+	WriteLockedPtr<ParameterT> operator->;
+	// optional: (useful for synchronized vector/map operations)
+	void applyRW(std::function<void(ParameterT& value)>);
+	void applyRO(std::function<void(const ParameterT& value)>);
+
+	// Copy-set/Copy-get value using read/write lock internally
+	ParameterT& operator=(const ParameterT& value);
+	ParameterT operator() const;  // or:  'operator ParameterT() const;'
+
+	// Validity Callbacks (could also use rcl_interfaces::msg::SetParametersResult like in node_parameters.h).
+	// Callbacks are stored in ordered hash map, could also possibly return hash for deleting later on.
+	// Alternatively, return Parameter<ParameterT>& for method chaining
+	using ParameterValidityCallback = std::function<bool(const ParameterT& value, std::string& message)>;
+	Parameter<ParameterT>& warnIf(ParameterValidityCallback warn_if_callback);
+	Parameter<ParameterT>& errorIf(ParameterValidityCallback error_if_callback);  // or rejectIf()/failIf()..
+	// Predefined generic (type-specific?) conditions, i.e. 'NO_OVERRIDE', 'EMPTY', 'NOT_NORMALIZED'
+	enum ParameterCondition
+	{
+		...
+	};
+	Parameter<ParameterT>& warnIf(ParameterCondition condition);
+	Parameter<ParameterT>& errorIf(ParameterCondition condition);
```

Comment:
Are you convinced it is actually useful to have a `warnIf` interface at all? I would argue that requests to set parameters either have to be valid or they are invalid. Allowing warnings here might support cases where the user knows better and sets parameters that trigger warnings (and they work). The result is a useless warning in an otherwise running system.

The other way around, if the warning actually indicates a failure, it should actually be an error anyway...

Reply:
It's probably up to the user. To me it's useful to warn that some feature is disabled or if some parameter doesn't have an override even though the default is valid. Sometimes warnings are just error cases that can be handled. In the end this API would allow making warnings errors and vice-versa without having to refactor custom error handling somewhere else in the code.